rfcs/006-delia-replication.md

```diff
@@ -0,0 +1,94 @@
+
+2. **Upload File:**  
+   For each deal CID returned from the proposal call, the client uploads the file via an HTTP PUT request to the corresponding endpoint (e.g. `/upload/{dealCid}`).
+
+3. **Publish Deal:**  
+   After file upload, the client:
+   - Encodes the deal proposal via the storage provider server’s HTTP endpoint (e.g. `/encode_proposal`).
+   - Uses the Polkadot extension to sign the encoded proposal.
+   - Publishes the deal via a JSON‑RPC call (e.g. `v0_publish_deal`).
+
+### Provider Selection
+
+Currently, the FE allows the user to select one storage provider from a list. With the multi-deal approach for replication, we will change that to let Delia automatically pick 3 different storage providers at random.
+
+### Storage Provider Considerations
+
+- **Server-Side Processing:**  
+  The storage provider server’s RPC endpoints remain unchanged. Since each storage provider runs its own server, the same client can submit three proposals without conflict.
+  
+- **On-Server Validation:**  
+  The storage provider’s automated acceptance parameters (as per RFC‑005) will determine which deals are accepted.
+
+### UX and Feedback
+
+- **Status Aggregation:**  
+  The FE should display the status of all three proposals (either aggregated or individually) so that the user is aware of the overall replication progress.
+- **Error Handling:**  
+  If one proposal fails while the others succeed, the client should clearly report the error and support a retry mechanism where it retries the failed storage provider 3 times, if all of them fail, Delia selects another storage provider at random, and repeats the process, aiming to ensure that the data is stored by 3 storage providers.
+
+### Documentation
+
+- Update the Delia documentation to reflect the new multi-deal submission process with an emphasis on replication.
+- Include examples showing how a client initiates three simultaneous proposals to achieve redundancy.
+- Document any new UI changes related to replication.
+
+## Implementation Details
+
+- **Looping Mechanism:**  
+  Modify the existing deal proposal code to iterate three times. Each iteration constructs the same deal proposal and calls the RPC endpoint.
+  
+- **Parallel Execution:**  
+  Use async patterns (e.g., `Promise.all`) to issue the three RPC calls concurrently.
+  
+- **Signature & Publishing:**  
+  The encoding and signing process remains unchanged; each deal proposal is individually signed and published.
+  
+- **Error Handling:**  
+  Aggregate results from the three RPC calls. If there is a failed upload attempt, Delia will retry 3 times (for each failed provider), if all 3 are unsuccessful it will select another storage provider at random and repeat this, with the goal of ensuring that the data is stored by at least 3 different storage providers.
+
+## Future work
+There will be work needed to make data retrieval be "replication-aware".
```

Comment:
Not relevant to this PR. A proper future work here would be "implement N-ary replication instead of forcing 3 copies"